From 2ab24d51b7f1d902cee501a2838c7a66dccb1808 Mon Sep 17 00:00:00 2001
From: QBos07 <qubos@outlook.de>
Date: Sat, 24 May 2025 13:53:10 +0000
Subject: [PATCH] SH: GAS: Treat BFD_RELOC_8/16/32 all the same way

While they should only diffrentiate between there sizes, 16 and 32 where
treated pretty diffrently from 8 wich resulted in the zeroing of .word
or .long jumptables when relaxing. Using .byte in these cases created it
correctly. This unifies all three and fixes such jumptabels.
---
 gas/config/tc-sh.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/gas/config/tc-sh.c b/gas/config/tc-sh.c
index db11939a350..2812228f567 100644
--- a/gas/config/tc-sh.c
+++ b/gas/config/tc-sh.c
@@ -3376,7 +3376,7 @@ apply_full_field_fix (fixS *fixP, char *buf, bfd_vma val, int size)
 {
   reloc_howto_type *howto;
 
-  if (fixP->fx_addsy != NULL || fixP->fx_pcrel)
+  if ((fixP->fx_addsy != NULL || fixP->fx_pcrel) && !SWITCH_TABLE (fixP))
     {
       howto = bfd_reloc_type_lookup (stdoutput, fixP->fx_r_type);
       if (howto && !howto->partial_inplace)
@@ -3533,7 +3533,6 @@ md_apply_fix (fixS *fixP, valueT *valP, segT seg ATTRIBUTE_UNUSED)
       *buf = val >> 2;
       break;
 
-    case BFD_RELOC_8:
     case BFD_RELOC_SH_IMM8:
       /* Sometimes the 8 bit value is sign extended (e.g., add) and
          sometimes it is not (e.g., and).  We permit any 8 bit value.
@@ -3541,7 +3540,7 @@ md_apply_fix (fixS *fixP, valueT *valP, segT seg ATTRIBUTE_UNUSED)
          reasonable looking assembly code, such as ``and -0x1,r0''.  */
       max = 0xff;
       min = -0xff;
-      *buf++ = val;
+      *buf = val;
       break;
 
     case BFD_RELOC_SH_PCRELIMM8BY4:
@@ -3603,13 +3602,19 @@ md_apply_fix (fixS *fixP, valueT *valP, segT seg ATTRIBUTE_UNUSED)
       buf[highbyte] |= (val >> 8) & 0xf;
       break;
 
-    case BFD_RELOC_32:
     case BFD_RELOC_32_PCREL:
       apply_full_field_fix (fixP, buf, val, 4);
       break;
 
+    case BFD_RELOC_8:
     case BFD_RELOC_16:
-      apply_full_field_fix (fixP, buf, val, 2);
+    case BFD_RELOC_32:
+      if (fixP->fx_size < sizeof(long))
+      {
+	max = (1ul << (fixP->fx_size * 8)) - 1;
+	min = -max;
+      }
+      apply_full_field_fix (fixP, buf, val, fixP->fx_size);
       break;
 
     case BFD_RELOC_SH_USES:
-- 
2.49.0

